OFEP-provider-client-mapping

```diff
@@ -0,0 +1,33 @@
+## Provider to client mapping
+
+## State: APPROVED
+
+The goal of this OFEP is to describe a way that application authors can selectively determine which provider a given client uses. This means that they should be able to set a provider for a library that is different than what they use in their main app.
+
+## Background
+
+Today, we operate with [a single provider stored on a global singleton](https://github.com/open-feature/spec/blob/74c373e089ad77bf8cac84f3d93c00c945ff3a8a/specification/sections/01-flag-evaluation.md?plain=1#L25). Especially in large apps, which may have different provider needs, this is limiting. Teams could solve this in user-space with a sort of `MultiplexingProvider` which takes in N other providers. Sadly, this introduces latency (calling provider 1 before falling through to provider 2 isn't free) and waste. Additionally, when it comes to testing, the "there can be only one!" nature of global singletons makes things a giant pain.
+
+## Proposal
+
+To address that, we will allow application authors to provide mappings of client names to a provider of their choosing. This should allow a great degree of flexibility for app authors to choose the flag-config store that works for that particular use-case.
+
+### Example Implementation
+
+Authors may optionally use [a named client](https://github.com/open-feature/spec/blob/74c373e089ad77bf8cac84f3d93c00c945ff3a8a/specification/sections/01-flag-evaluation.md?plain=1#L60), which we already support. If they don't pick a name, it falls through to the configured "default" provider (or no provider if none are set).
+
+```java
+# example library code
+var client = OpenFeature.getInstance().getClient('my-sweet-library')
```

Comment:
Do we need to document about the SDK behaviour when no provider exists for the given client name?

Reply:
We will in the spec, but as mentioned.. it falls through to the default provider.

To be clear, 'my sweet library' is the name of the client, not the provider.

Reply:
Just to make sure that, I understand it properly, If I do `OpenFeature.getInstance().getClient('not_exists')`, you meant to say that it fallback to the default provider.  If so, it's better we let the application author to decide on the behaviour they want if the expected client is not available rather than we fall back to default provider. 

```
OpenFeature.getInstance().getClient() or OpenFeature.getInstance().getClient(OpenFeature.DEFAULT_PROVIDER) -> FlagDProvider
OpenFeature.getInstance().getClient('my-sweet-library') -> FileBasedProvider 
OpenFeature.getInstance().getClient('not_exists') -> null
``

Reply:
You're not quite getting what I'm saying. The `getClient()` function does not take a provider name. It takes an optional client name.  So to reuse your example:

```
OpenFeature.getInstance().getClient() -> flagD provider
OpenFeature.getInstance().getClient(OpenFeature.DEFAULT_PROVIDER) // this is nonsensical. It doesn't reference a provider
OpenFeature.getInstance().getClient('my-sweet-library') -> FileBasedProvider 
OpenFeature.getInstance().getClient('not_exists') -> flagD provider
```

Reply:
> OpenFeature.getInstance().getClient(OpenFeature.DEFAULT_PROVIDER) // this is nonsensical. It doesn't reference a provider

If I could set a provider with this (as per the example) OpenFeature.setProvider(OpenFeature.DEFAULT_PROVIDER, new FlagDProvider()) where OpenFeature.DEFAULT_PROVIDER is String constant, then why it is nonsensical get the client with the same constant?

> OpenFeature.getInstance().getClient('my-sweet-library') -> FileBasedProvider 

I understand that we're mapping the client name with the provider name which is used here `OpenFeature.setProvider('my-sweet-library', new FileBasedProvider('path/to/myfile'))` to get the client with specific feature config store

Is there a way I could get `FileBasedProvider` as feature config store with `OpenFeature.getInstance().getClient('not_exists')`. If not, then my client name should be always `my-sweet-library` to get `FileBasedProvider` as my feature config store. 

Hope I'm clear to you.

Reply:
> > OpenFeature.getInstance().getClient(OpenFeature.DEFAULT_PROVIDER) // this is nonsensical. It doesn't reference a provider
>
> If I could set a provider with this (as per the example) OpenFeature.setProvider(OpenFeature.DEFAULT_PROVIDER, new FlagDProvider()) where OpenFeature.DEFAULT_PROVIDER is String constant, then why it is nonsensical to get the client with the same constant?

It is syntactically valid to pass in the DEFAULT_PROVIDER constant, however it's confusing the nature of things. What you pass to a client is the name of the client, not an indication of which providers to use.

---



> >  OpenFeature.getInstance().getClient('not_exists') -> flagD provider
>
> Is there a way I could get FileBasedProvider as feature config store with OpenFeature.getInstance().getClient('not_exists'). If not, then my client name should be always my-sweet-library to get FileBasedProvider as my feature config store. Is that correct?

This is similarly confusing things. `not_exists` returns the configured default provider. This isn't because `not_exists` is some special string, but because we didn't configure a mapping of client name to a different provider.

While it really makes things confusing for this example, an application author could configure `setProvider('not_exists', myFileProvider)` and it would return the file provider.  The client name to choose is the decision of people using clients (library authors and application authors). How to actually map those names to feature providers is solely the decision of the application author when they configure things.

Reply:
@justinabrahms I think I like this proposal, but I had similar confusion to @thiyagu06 initially trying to grok this proposal, though that could be because my proposal was to create a provider map.

I wonder if there's any tiny changes we can make to avoid this confusion. I also wonder if we would want to support identifying more than one client name to associate with the provider in a single call.

Reply:
@toddbaert , Yes, the proposal would address my issue. Still I have concerns which may be related to implementation of SDK.

1. when I call `setProvider('my-sweet-library', myFileProvider)`, it seems that we are mixing up the `client_name` with provider instances. This is the reason for all my confusion. I'm not quite sure how easy to understand this behaviour for the wider audience as well.

I'm also wondering can we reuse the  provider name from the metadata https://github.com/open-feature/java-sdk/blob/main/src/main/java/dev/openfeature/sdk/FeatureProvider.java#L10. so when registering the provider we can resolve the name from metadata object to create a map of providers and then provide an API like below to get the client for specific provider `OpenFeature.getInstance().getClientForProvider('flagd')` ( we can come up with better name 🤔 )

You can refer this to get some inspiration on how open telemetry is solving similar issue. https://github.com/open-telemetry/oteps/blob/main/text/0016-named-tracers.md#explanation

2. If I have only one provider in my application, how does the SDK will looks like? Are we going to force the developer to set the default provider like this `OpenFeature.setProvider(OpenFeature.DEFAULT_PROVIDER, new FlagDProvider())`  or  if the name is not provided while registering the provider  like this `OpenFeature.setProvider(new FlagDProvider())`, then the `FlagDProvider` will become a default provider?

Reply:
> 1. when I call `setProvider('my-sweet-library', myFileProvider)`, it seems that we are mixing up the `client_name` with provider instances. This is the reason for all my confusion. I'm not quite sure how easy to understand this behaviour for the wider audience as well.

I share this concern because I had the same exact confusion earlier. On the other hand, I think this is better than forcing application authors to know about provider mappings. I think that is a concern that @moredip and @justinabrahms both have, and I agree with it.

Part of it could be naming? Improved naming can help communicate concepts a lot. Perhaps instead of client name, we want to find a better name... like `scope`? cc @justinabrahms @moredip 

> 2. If I have only one provider in my application, how does the SDK will looks like? Are we going to force the developer to set the default provider like this `OpenFeature.setProvider(OpenFeature.DEFAULT_PROVIDER, new FlagDProvider())`  or  if the name is not provided while registering the provider  like this `OpenFeature.setProvider(new FlagDProvider())`, then the `FlagDProvider` will become a default provider?

I think if you don't mention a client name (maybe `scope` moving forward) you the registered provider stays global and the current behavior persists.

Reply:
Thanks @toddbaert. This make sense to me now 👍